docs/rfc/0001_ctrlkit.md

```diff
@@ -0,0 +1,325 @@
+
+Function  | Effect 
+:---: | :---: 
+RequeueImmediately | Interrupts the workflow and requeue the request immediately.
+RequeueAfter | Interrupts the workflow and requeue the request after given time.
+RequeueIfError | Interrupts the workflow and requeue the request into queue if there's an error.
+NoRequeue | No effect.
+Exit | Equivalent to `RequeueIfError(ErrExit)`, where `ErrExit` is a pre-defined error to exit the workflow without reason.
+
+**Rules of result join**
+
+* `nil` join `err` = `err`
+* `err1` join `err2` = `multierr{err1, err2}`
+* `Result.Requeue`s are OR-ed.
+* `Result.RequeueAfter`, we use the shortest non-zero duration if there's any, or 0 value otherwise.
+
+These rules ensure that the interruption always happens if any action wants it and never happens later than expected. 
+
+### **Controller**
+
+Here's a simple example of how to implement a controller with the framework:
+
+```go
+type CronJobController struct {
+	client.Client
+	logr.Logger
+}
+
+func (c *CronJobController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := c.Logger.WithValues("cronjob", request)
+
+	// Get CronJob object with client.
+	var cronJob apiv1.CronJob
+	if err := c.Client.Get(ctx, request.NamespacedName, &cronJob); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("object not found, skip")
+			return ctrlkit.NoRequeue()
+		}
+		logger.Error(err, "unable to get object")
+		return ctrlkit.RequeueIfError(err)
+	}
+
+	// Build state and impl for controller manager.
+	state := manager.NewCronJobControllerManagerState(c.Client, cronJob.DeepCopy())
+	impl := manager.NewCronJobControllerManagerImpl(c.Client, cronJob.DeepCopy())
+	mgr := manager.NewCronJobControllerManager(state, impl, logger)
+
+	// Always update the status after actions have run.
+	defer mgr.UpdateCronJobStatus()
+
+	// Assemble the actions and run.
+	return ctrlkit.IgnoreExit(
+		// Run these actions and doesn't care the order, and join the results.
+		ctrlkit.Join(
+			// Update the status of CronJob as always.
+			mgr.ListActiveJobsAndUpdateStatus(),
+			// Clean the old completed/failed jobs accroding to the limits.
+			mgr.CleanUpOldJobsExceedsHistoryLimits(),
+			// Try to run the next scheduled job when not suspended, otherwise do nothing.
+			ctrlkit.If(cronJob.Spec.Suspend == nil || *cronJob.Spec.Suspend, mgr.RunNextScheduledJob()),
+		).Run(ctx),
+	)
+}
+```
+
+### **Not mentioned or not related**
+
+1. How to trace the actions and workflow? We could achieve this by logging before and after each action run.
+2. Problems supposed to be considered while implementing the controller:
+	* When and how to observe reality? Is cache necessary? Will cache introduce additional issues?
+	* When to update the target's status? How do we deal with the conflicts?
+
+# **Drawbacks**
+
+1. The current design is tightly tied to the `controller-runtime`.
+2. You have to regenerate the stub each time the DSL doc changes.
```

Comment:
Is it costly to maintain the CLI as well?

Reply:
Currently, it's fine. Not sure if there will be many new features.